src/main/scala/com/wavesplatform/state2/BlockchainUpdaterImpl.scala

```diff
@@ -120,7 +120,25 @@ class BlockchainUpdaterImpl private(persisted: StateWriter with StateReader,
       bottomMemoryDiff.set(topMemoryDiff())
       topMemoryDiff.set(BlockDiff.empty)
     }
-    (ngState() match {
+
+    val height = historyWriter.height()
+    val notImplementedFeatures = featureProvider.activatedFeatures(height).diff(BlockchainFeatures.implemented)
+
+    for {
+      _ <- Either.cond(settings.featuresSettings.autoShutdownOnUnsupportedFeature && notImplementedFeatures.isEmpty, (),
+        GenericError(s"UNIMPLEMENTED ${displayFeatures(notImplementedFeatures)} ACTIVATED ON BLOCKCHAIN, UPDATE THE NODE IMMEDIATELY"))
+      ng <- processNg(block)
+      (newBlockDiff, discarded) = ng
+    } yield {
+      ngState.set(Some(NgState(block, newBlockDiff, 0L, featuresApprovedWithBlock(block))))
+      historyReader.lastBlockId().foreach(lastBlockId.onNext)
+      log.info(s"Block ${block.uniqueId} -> ${trim(block.reference)} appended. New height: ${height + 1}, transactions: ${block.transactionData.size})")
+      discarded
+    }
+  }
+
+  private def processNg(block: Block): Either[ValidationError, (BlockDiff, Seq[Transaction])] = write { implicit l =>
```

Comment:
Something like consensusValidation makes more sense.
Just a suggesstion, you can make it `lazy val blockConsensusValidation = ...`:

```
lazy val blockConsensusValidation = ngState() match {
   ...
}
 for {
      _ <- Either.cond(settings.featuresSettings.autoShutdownOnUnsupportedFeature && notImplementedFeatures.isEmpty, (), GenericError(s"UNIMPLEMENTED ${displayFeatures(notImplementedFeatures)} ACTIVATED ON BLOCKCHAIN, UPDATE THE NODE IMMEDIATELY"))
      newBlockDiffAndDiscardedTxs<- blockConsensusValidation 
      (newBlockDiff, discarded) = newBlockDiffAndDiscardedTxs
    }  ...
```